python-backend/main.py

```diff
@@ -145,3 +63,69 @@ async def analyze(request: Request):
+    chart coordinates to be rendered by the frontend
+    """
+    await websocket.accept()
+    while True:
+        try:
+            data = Input(**(await websocket.receive_json()))
+            logger.debug('received data. parsing...')
+            iterations = data.i
+            (a_func, a, b_func, b, symbol) = parse(data)
+
+            if len(data.symbol) > 0:
+                logger.debug('checking for convergence...')
+                is_convergent = assess_convergence(laurent(a, b, symbol), symbol)
+                logger.debug('convergence result: {}'.format(is_convergent))
+                # short circuit if the preliminary test fails
+                await websocket.send_json({"is_convergent": is_convergent})
+
+                if is_convergent is False:
+                    logger.debug(f"values do not converge. closing socket.")
+                    await websocket.close()
+
+            # per the link below, "when bi = 1 (the partial numerator) for all i the expression is called a
+            # simple continued fraction" b here is the partial numerator (a and b are often used interchangeably
+            # which may lead to confusion when there is no associated image of the continued fraction)
+            # https://en.wikipedia.org/wiki/Continued_fraction#Basic_formula
+            if data.b == "1":
+                (values, num_values, denom_values) = simple_computed_values(a_func, iterations=iterations)
+            else:
+                # generalized continued fractions, where the partial numerator has its own formula and is not
+                # equal to 1 https://en.wikipedia.org/wiki/Generalized_continued_fraction
+                (values, num_values, denom_values) = generalized_computed_values(a_func, b_func,
+                                                                                 iterations=iterations)
+
+            limit = values[next(reversed(values))]
+            logger.debug(f"last convergent / limit: {limit}")
+            await websocket.send_json({"limit": "Infinity" if type(limit) is Infinity else str(limit)})
+
+            computed_values = db.identify(values=[str(limit)], wide_search=True)
+            for m in computed_values:
+                logger.debug(f"identify returned: {m}")
+
+            await websocket.send_json(
+                {"converges_to": json.dumps(str(computed_values[0] if len(computed_values) > 0 else None))})
+
+            await chart_coordinates(values,
+                                    num_values,
+                                    denom_values,
+                                    limit,
+                                    iterations=iterations,
+                                    websocket=websocket)
+
+        except WebSocketDisconnect:
+            logger.debug("Websocket disconnected")
+            break
+        except WebSocketException:
+            logger.debug("Websocket exception")
+            break
+
+        await websocket.close()
```

Comment:
Should this line be indented at the same level as `websocket.accept()`? 

If not, there may be scoping issues between the `try` clause and the `while True:` clause.

Reply:
Specifically, I noticed that `close` is not called when an exception is encountered.